controllers/flinkcluster_observer.go

```diff
@@ -135,35 +136,78 @@ func (observer *_ClusterStateObserver) observe(
 	}
 
 	// (Optional) job.
-	var observedJob *batchv1.Job
-	if observedState.cluster != nil && observedState.cluster.Spec.JobSpec != nil {
-		observedJob = new(batchv1.Job)
-		err = observer.observeJob(observedJob)
-		if err != nil {
-			if client.IgnoreNotFound(err) != nil {
-				log.Error(err, "Failed to get job")
-				return err
-			} else {
-				log.Info("Observed job", "state", "nil")
-				observedJob = nil
-			}
+	err = observer.observeJob(observedState)
+
+	return err
+}
+
+func (observer *_ClusterStateObserver) observeJob(
+	observedState *_ObservedClusterState) error {
+	var err error
+	var log = observer.log
+
+	// Either the cluster has been deleted or it is a session cluster.
+	if observedState.cluster == nil ||
+		observedState.cluster.Spec.JobSpec == nil {
+		return nil
+	}
+
+	// Job resource.
+	var observedJob = new(batchv1.Job)
+	err = observer.observeJobResource(observedJob)
+	if err != nil {
+		if client.IgnoreNotFound(err) != nil {
+			log.Error(err, "Failed to get job")
+			return err
 		} else {
-			log.Info("Observed job", "state", *observedJob)
-			observedState.job = observedJob
+			log.Info("Observed job", "state", "nil")
+			observedJob = nil
 		}
+	} else {
+		log.Info("Observed job", "state", *observedJob)
+		observedState.job = observedJob
 	}
 
-	// (Optional) get Flink job ID.
-	if observedCluster != nil && observedJob != nil && observedJmService != nil {
+	// Job pod.
+	var observedJobPods = new(corev1.PodList)
+	err = observer.observeJobPods(observedJobPods)
+	if err != nil {
+		if client.IgnoreNotFound(err) != nil {
+			log.Error(err, "Failed to get job pods")
+			return err
+		} else {
+			log.Info("Observed job pods", "pods", "nil")
+			observedJobPods = nil
+		}
+	} else {
+		log.Info("Observed job pods", "pods", observedJobPods.Items)
+		if len(observedJobPods.Items) == 1 {
```

Comment:
Should we log something here if no job pod is found at all?

Reply:
The previous line will log empty list if not job pod is found, and it is a normal condition, not an error.